src/core/ext/client_channel/subchannel.c

```diff
@@ -788,42 +789,39 @@ grpc_call_stack *grpc_subchannel_call_get_call_stack(
   return SUBCHANNEL_CALL_TO_CALL_STACK(subchannel_call);
 }
 
-static void grpc_uri_to_sockaddr(const char *uri_str,
-                                 grpc_resolved_address *addr) {
-  grpc_uri *uri = grpc_uri_parse(uri_str, 0 /* suppress_errors */);
-  GPR_ASSERT(uri != NULL);
-  if (strcmp(uri->scheme, "ipv4") == 0) {
-    GPR_ASSERT(parse_ipv4(uri, addr));
-  } else if (strcmp(uri->scheme, "ipv6") == 0) {
-    GPR_ASSERT(parse_ipv6(uri, addr));
-  } else {
-    GPR_ASSERT(parse_unix(uri, addr));
-  }
-  grpc_uri_destroy(uri);
-}
-
-void grpc_get_subchannel_address_arg(const grpc_channel_args *args,
+void grpc_get_subchannel_address_arg(grpc_exec_ctx *exec_ctx,
+                                     const grpc_channel_args *args,
                                      grpc_resolved_address *addr) {
-  const char *addr_uri_str = grpc_get_subchannel_address_uri_arg(args);
-  memset(addr, 0, sizeof(*addr));
-  if (*addr_uri_str != '\0') {
-    grpc_uri_to_sockaddr(addr_uri_str, addr);
-  }
-}
-
-const char *grpc_get_subchannel_address_uri_arg(const grpc_channel_args *args) {
   const grpc_arg *addr_arg =
       grpc_channel_args_find(args, GRPC_ARG_SUBCHANNEL_ADDRESS);
   GPR_ASSERT(addr_arg != NULL);  // Should have been set by LB policy.
   GPR_ASSERT(addr_arg->type == GRPC_ARG_STRING);
-  return addr_arg->value.string;
+  memset(addr, 0, sizeof(*addr));
+  if (*addr_arg->value.string != '\0') {
+    grpc_slice addr_slice =
+        grpc_base64_decode(exec_ctx, addr_arg->value.string, 0 /* url_safe */);
+    GPR_ASSERT(GRPC_SLICE_LENGTH(addr_slice) <= GRPC_MAX_SOCKADDR_SIZE);
+    memcpy(addr->addr, (const char *)GRPC_SLICE_START_PTR(addr_slice),
+           GRPC_SLICE_LENGTH(addr_slice));
+    addr->len = GRPC_SLICE_LENGTH(addr_slice);
+    grpc_slice_unref(addr_slice);
+  }
+}
+
+char *grpc_get_subchannel_address_uri_arg(grpc_exec_ctx *exec_ctx,
```

Comment:
There's probably no need to provide this wrapper function anymore if the value is not actually going to be a URI.  Instead, I would just change the one caller of this to convert the value to a URI itself.